pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/RealtimeSegmentAssignment.java

```diff
@@ -345,6 +380,11 @@ private Map<String, Map<String, String>> reassignSegments(String instancePartiti
 
         newAssignment = SegmentAssignmentUtils
             .rebalanceReplicaGroupBasedTable(currentAssignment, instancePartitions, partitionGroupIdToSegmentsMap);
+        for (String segName : uploadedSegmentsWithNoPartitionMetadata) {
```

Comment:
We should combine the unpartitioned segments into the `partitionGroupIdToSegmentsMap` (assign a partition id to them). I'd suggest simply using the `Math.abs(segmentName.hashCode() % numPartitions)` as the partition group id because it is deterministic. We need to keep it deterministic so that during multiple rebalances, it won't switch partition which will force it to move around different servers